Vault segfaults when a custom plugin stores a nil StorageEntry

**Describe the bug**

I'm writing a custom auth backend for Vault, and I accidentally called `req.Storage.Put()` with a `nil` value. This caused vault itself (not my plugin) to segfault:

```==> Vault server configuration:

             Api Address: http://127.0.0.1:8200
                     Cgo: disabled
         Cluster Address: https://127.0.0.1:8201
              Listener 1: tcp (addr: "127.0.0.1:8200", cluster address: "127.0.0.1:8201", tls: "disabled")
               Log Level: info
                   Mlock: supported: false, enabled: false
                 Storage: inmem
                 Version: Vault v0.10.1
             Version Sha: 756fdc4587350daf1c65b93647b2cc31a6f119cd

WARNING! dev mode is enabled! In this mode, Vault runs entirely in-memory
and starts unsealed with a single unseal key. The root token is already
authenticated to the CLI, so you can immediately begin using Vault.

You may need to set the following environment variable:

    $ export VAULT_ADDR='http://127.0.0.1:8200'

The unseal key and root token are displayed below in case you want to
seal/unseal the Vault or re-authenticate.

Unseal Key: jJ5PdT0mpXKHe11sfKtZOK/2NixQm9baO8ldBqW/ca8=
Root Token: root

Development mode should NOT be used in production installations!

==> Vault server started! Log data will stream in below:

2018-06-20T13:13:50.064+0200 [INFO ] core: security barrier not initialized
2018-06-20T13:13:50.064+0200 [INFO ] core: security barrier initialized: shares=1 threshold=1
2018-06-20T13:13:50.065+0200 [INFO ] core: post-unseal setup starting
2018-06-20T13:13:50.079+0200 [INFO ] core: loaded wrapping token key
2018-06-20T13:13:50.079+0200 [INFO ] core: successfully setup plugin catalog: plugin-directory=/Users/jerith/.gostuff/src/github.com/hashicorp/vault-auth-plugin-example
2018-06-20T13:13:50.079+0200 [INFO ] core: no mounts; adding default mount table
2018-06-20T13:13:50.079+0200 [INFO ] core: successfully mounted backend: type=kv path=secret/
2018-06-20T13:13:50.079+0200 [INFO ] core: successfully mounted backend: type=cubbyhole path=cubbyhole/
2018-06-20T13:13:50.080+0200 [INFO ] core: successfully mounted backend: type=system path=sys/
2018-06-20T13:13:50.080+0200 [INFO ] core: successfully mounted backend: type=identity path=identity/
2018-06-20T13:13:50.082+0200 [INFO ] core: restoring leases
2018-06-20T13:13:50.082+0200 [INFO ] rollback: starting rollback manager
2018-06-20T13:13:50.082+0200 [INFO ] expiration: lease restore complete
2018-06-20T13:13:50.082+0200 [INFO ] identity: entities restored
2018-06-20T13:13:50.082+0200 [INFO ] identity: groups restored
2018-06-20T13:13:50.082+0200 [INFO ] core: post-unseal setup complete
2018-06-20T13:13:50.083+0200 [INFO ] core: root token generated
2018-06-20T13:13:50.083+0200 [INFO ] core: pre-seal teardown starting
2018-06-20T13:13:50.083+0200 [INFO ] core: cluster listeners not running
2018-06-20T13:13:50.083+0200 [INFO ] rollback: stopping rollback manager
2018-06-20T13:13:50.083+0200 [INFO ] core: pre-seal teardown complete
2018-06-20T13:13:50.083+0200 [INFO ] core: vault is unsealed
2018-06-20T13:13:50.083+0200 [INFO ] core: post-unseal setup starting
2018-06-20T13:13:50.083+0200 [INFO ] core: loaded wrapping token key
2018-06-20T13:13:50.083+0200 [INFO ] core: successfully setup plugin catalog: plugin-directory=/Users/jerith/.gostuff/src/github.com/hashicorp/vault-auth-plugin-example
2018-06-20T13:13:50.083+0200 [INFO ] core: successfully mounted backend: type=kv path=secret/
2018-06-20T13:13:50.083+0200 [INFO ] core: successfully mounted backend: type=system path=sys/
2018-06-20T13:13:50.083+0200 [INFO ] core: successfully mounted backend: type=identity path=identity/
2018-06-20T13:13:50.083+0200 [INFO ] core: successfully mounted backend: type=cubbyhole path=cubbyhole/
2018-06-20T13:13:50.084+0200 [INFO ] core: restoring leases
2018-06-20T13:13:50.084+0200 [INFO ] rollback: starting rollback manager
2018-06-20T13:13:50.084+0200 [INFO ] expiration: lease restore complete
2018-06-20T13:13:50.084+0200 [INFO ] identity: entities restored
2018-06-20T13:13:50.084+0200 [INFO ] identity: groups restored
2018-06-20T13:13:50.084+0200 [INFO ] core: post-unseal setup complete
2018-06-20T13:13:50.088+0200 [INFO ] core: mount tuning of options: path=secret/ options=map[version:2]
2018-06-20T13:13:50.088+0200 [INFO ] secrets.kv.kv_abdb7de7: collecting keys to upgrade
2018-06-20T13:13:50.088+0200 [INFO ] secrets.kv.kv_abdb7de7: done collecting keys: num_keys=1
2018-06-20T13:13:50.088+0200 [INFO ] secrets.kv.kv_abdb7de7: upgrading keys finished
2018-06-20T13:13:57.616+0200 [WARN ] auth.plugin.auth_plugin_078d7e18.example-auth-plugin: error closing client during Kill: metadata=true err="rpc error: code = Canceled desc = grpc: the client connection is closing"
2018-06-20T13:13:57.619+0200 [INFO ] core: enabled credential backend: path=example/ type=plugin
2018-06-20T13:14:38.060+0200 [INFO ] expiration: revoked lease: lease_id=auth/example/login/55a005e9dd21cab9cca50aa140641bd9483302bc
2018-06-20T13:14:59.048+0200 [WARN ] auth.plugin.auth_plugin_078d7e18.example-auth-plugin: error closing client during Kill: err="rpc error: code = Canceled desc = grpc: the client connection is closing"
2018-06-20T13:14:59.051+0200 [WARN ] auth.plugin.auth_plugin_078d7e18.example-auth-plugin: error closing client during Kill: err="rpc error: code = Canceled desc = grpc: the client connection is closing"
2018-06-20T13:14:59.051+0200 [INFO ] core: disabled credential backend: path=example/
2018-06-20T13:15:00.647+0200 [WARN ] auth.plugin.auth_plugin_71820bc9.example-auth-plugin: error closing client during Kill: metadata=true err="rpc error: code = Canceled desc = grpc: the client connection is closing"
2018-06-20T13:15:00.649+0200 [INFO ] core: enabled credential backend: path=example/ type=plugin
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x2571f16]

goroutine 384 [running]:
github.com/hashicorp/vault/vault.(*BarrierView).Put(0xc4200d5a90, 0x30f6300, 0xc42029fa70, 0x0, 0x2b64d40, 0x0)
	/private/tmp/vault-20180425-78417-qvq6c1/src/github.com/hashicorp/vault/vault/barrier_view.go:92 +0x26
github.com/hashicorp/vault/logical/plugin.(*GRPCStorageServer).Put(0xc420aee0f0, 0x30f6300, 0xc42029fa70, 0xc4200d2030, 0xc420aee0f0, 0x0, 0xc420a50a70)
	/private/tmp/vault-20180425-78417-qvq6c1/src/github.com/hashicorp/vault/logical/plugin/grpc_storage.go:99 +0x63
github.com/hashicorp/vault/logical/plugin/pb._Storage_Put_Handler(0x2b64d40, 0xc420aee0f0, 0x30f6300, 0xc42029fa70, 0xc42048a150, 0x0, 0x0, 0x0, 0x3f259091ee22e143, 0x6b5d36450bbcb5b0)
	/private/tmp/vault-20180425-78417-qvq6c1/src/github.com/hashicorp/vault/logical/plugin/pb/backend.pb.go:1932 +0x241
github.com/hashicorp/vault/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc4202c66e0, 0x30fe480, 0xc420a97200, 0xc420678000, 0xc4201f0ab0, 0x40ff590, 0x0, 0x0, 0x0)
	/private/tmp/vault-20180425-78417-qvq6c1/src/github.com/hashicorp/vault/vendor/google.golang.org/grpc/server.go:920 +0x848
github.com/hashicorp/vault/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4202c66e0, 0x30fe480, 0xc420a97200, 0xc420678000, 0x0)
	/private/tmp/vault-20180425-78417-qvq6c1/src/github.com/hashicorp/vault/vendor/google.golang.org/grpc/server.go:1142 +0x1318
github.com/hashicorp/vault/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4208c4070, 0xc4202c66e0, 0x30fe480, 0xc420a97200, 0xc420678000)
	/private/tmp/vault-20180425-78417-qvq6c1/src/github.com/hashicorp/vault/vendor/google.golang.org/grpc/server.go:637 +0x9f
created by github.com/hashicorp/vault/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/private/tmp/vault-20180425-78417-qvq6c1/src/github.com/hashicorp/vault/vendor/google.golang.org/grpc/server.go:635 +0xa1
```

**To Reproduce**

1. Start with the example plugin from https://github.com/hashicorp/vault-auth-plugin-example and apply the following patch:

```diff
diff --git a/main.go b/main.go
index f156223..3ea8742 100644
--- a/main.go
+++ b/main.go
@@ -69,13 +69,15 @@ func Backend(c *logical.BackendConfig) *backend {
        return &b
 }

-func (b *backend) pathAuthLogin(_ context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+func (b *backend) pathAuthLogin(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
        password := d.Get("password").(string)

        if subtle.ConstantTimeCompare([]byte(password), []byte("super-secret-password")) != 1 {
                return nil, logical.ErrPermissionDenied
        }

+       req.Storage.Put(ctx, nil)
+
        // Compose the response
        return &logical.Response{
                Auth: &logical.Auth{
```

2. Run vault with a config that specifies an appropriate `plugin_directory`. (I used the repo root, full config file below.)

3. Run the following commands to build and enable the plugin:
```
go build -o=vault-auth-plugin-example
SHASUM=$(shasum -a 256 vault-auth-plugin-example | cut -d' ' -f1)
vault write sys/plugins/catalog/example-auth-plugin sha_256=$SHASUM command=vault-auth-plugin-example
vault auth enable -path=example -plugin-name=example-auth-plugin plugin
```

4. Run the following command to invoke the broken plugin code:
```
vault write auth/example/login password=super-secret-password
```

**Expected behavior**

I would expect an error in the plugin (perhaps crashing the plugin) and an error response from the API call that triggered the problem, not Vault itself crashing.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 0.10.1
* Vault CLI Version (retrieve with `vault version`): Vault v0.10.1 ('756fdc4587350daf1c65b93647b2cc31a6f119cd')
* Server Operating System/Architecture: OSX 10.12

Vault server configuration file(s):

```hcl
plugin_directory = "/Users/jerith/.gostuff/src/github.com/hashicorp/vault-auth-plugin-example"
```

**Additional context**

I'm running Vault from homebrew, but the plugin has the same version (0.10.1) as a dependency.

--- a/vault/barrier_view.go
+++ b/vault/barrier_view.go
@@ -89,6 +89,10 @@ func (v *BarrierView) Get(ctx context.Context, key string) (*logical.StorageEntr
 
 // logical.Storage impl.
 func (v *BarrierView) Put(ctx context.Context, entry *logical.StorageEntry) error {
+	if entry == nil {
+		return errors.New("cannot write nil entry")
+	}
+
 	if err := v.sanityCheck(entry.Key); err != nil {
 		return err
 	}

